File: fboss/platform/bsp_tests/utils/gpio_utils.py
import re
from dataclasses import dataclass
from typing import List, Optional

from fboss.platform.bsp_tests.utils.cmd_utils import run_cmd


@dataclass
class GpioDetectResults:
    name: str
    label: str
    lines: int


@dataclass
class GpioInfoResult:
    name: str
    lineNum: int
    direction: str


def gpiodetect(label: str | None = None) -> list[GpioDetectResults]:
    ret: list[GpioDetectResults] = []
    results = run_cmd(["gpiodetect"]).stdout.decode()
    # format of gpiodetect output is:
    # <name> [<label>] (<num_lines> lines)
    for line in results.splitlines():
        match = re.match(r"(\w+)\s+\[(.*?)\]\s+\((.*?) lines\)", line)
        if match:
            name = match.group(1)
            label = match.group(2)
            num_lines = int(match.group(3))
            ret.append(GpioDetectResults(name, label, num_lines))
    if label is None:
        return ret
    else:
        return [x for x in ret if label == x.label]
    return ret


def gpioinfo(name: str) -> list[GpioInfoResult]:
    """
    the output format of gpioinfo changes significantly between
    versions, so instead of strictly parsing it, we only extract
    the values that we will want to check to avoid needing a
    lot of logic to deal with the differences.
    """
    ret: list[GpioInfoResult] = []
    results = run_cmd(["gpioinfo", name]).stdout.decode()
    for line in results.splitlines()[1:]:
        parts = line.split()
        lineNum = parts[1][:-1]  # remove ":" from line number
        name = parts[2]
        direction = "None"
        if "input" in parts:
            direction = "input"
        elif "output" in parts:
            direction = "output"
        ret.append(GpioInfoResult(name, lineNum, direction))

    return ret


def gpioget(name: str, line: int) -> int:
    ret = run_cmd(["gpioget", name, str(line)])
    if ret.returncode != 0:
        raise Exception(f"gpioget {name}-{line}failed with return code {ret}")
    return int(ret.stdout.decode())

File: fboss/platform/bsp_tests/utils/GpioUtils.cpp
#include "fboss/platform/bsp_tests/utils/GpioUtils.h"

#include <re2/re2.h>
#include <sstream>
#include <stdexcept>
#include <string>

#include <fmt/format.h>
#include <folly/logging/xlog.h>

#include "fboss/platform/helpers/PlatformUtils.h"

namespace facebook::fboss::platform::bsp_tests {

std::vector<GpioDetectResult> GpioUtils::gpiodetect(const std::string& label) {
  std::vector<GpioDetectResult> results;

  try {
    auto [exitStatus, output] = PlatformUtils().execCommand("gpiodetect");
    if (exitStatus != 0) {
      XLOG(ERR) << "gpiodetect command failed with exit status " << exitStatus;
      return results;
    }
    re2::RE2 pattern(R"((\w+)\s+\[(.*?)\]\s+\((.*?) lines\))");
    std::string deviceName, deviceLabel, numLinesStr;
    re2::StringPiece input(output);

    while (RE2::FindAndConsume(
        &input, pattern, &deviceName, &deviceLabel, &numLinesStr)) {
      int numLines = std::stoi(numLinesStr);

      // If label is provided, only add devices with matching label
      if (label.empty() || deviceLabel == label) {
        results.push_back({deviceName, deviceLabel, numLines});
      }
    }
  } catch (const std::exception& e) {
    XLOG(ERR) << "Error in gpiodetect: " << e.what();
  }

  return results;
}

std::vector<GpioInfoResult> GpioUtils::gpioinfo(const std::string& name) {
  std::vector<GpioInfoResult> results;

  try {
    auto [exitStatus, output] =
        PlatformUtils().execCommand(fmt::format("gpioinfo {}", name));
    if (exitStatus != 0) {
      XLOG(ERR) << "gpioinfo command failed with exit status " << exitStatus;
      return results;
    }
    std::istringstream stream(output);
    std::string line;

    // Skip the header line
    std::getline(stream, line);

    while (std::getline(stream, line)) {
      std::istringstream lineStream(line);
      std::string token;
      // Skip the first token (line name)
      lineStream >> token;
      // Get line number (remove the trailing colon)
      lineStream >> token;
      int lineNum = std::stoi(token.substr(0, token.size() - 1));
      // Get line name
      lineStream >> token;
      std::string lineName = token;
      // Determine direction
      std::string direction = "None";
      while (lineStream >> token) {
        if (token == "input") {
          direction = "input";
          break;
        } else if (token == "output") {
          direction = "output";
          break;
        }
      }
      results.push_back({lineName, lineNum, direction});
    }
  } catch (const std::exception& e) {
    XLOG(ERR) << "Error in gpioinfo: " << e.what();
  }

  return results;
}

int GpioUtils::gpioget(const std::string& name, int line) {
  try {
    auto [exitStatus, output] =
        PlatformUtils().execCommand(fmt::format("gpioget {} {}", name, line));
    if (exitStatus != 0) {
      throw std::runtime_error(
          fmt::format(
              "gpioget command failed with exit status {}", exitStatus));
    }
    return std::stoi(output);
  } catch (const std::exception& e) {
    XLOG(ERR) << "Error in gpioget: " << e.what();
    throw std::runtime_error(
        fmt::format(
            "Failed to get GPIO value for {}-{}: {}", name, line, e.what()));
  }
}

} // namespace facebook::fboss::platform::bsp_tests
